src/service.rs

```diff
@@ -568,66 +614,166 @@ impl IndexerRpc for IndexerRpcImpl {
-                                            CellType::Output => indexer::CellType::Output,
-                                        },
-                                    )
-                                    .into_vec(),
+                                .into_vec(),
+                            )
+                            .expect("get TxLockScript should be OK")
+                            .is_some(),
+                        ScriptType::Type => snapshot
+                            .get(
+                                Key::TxTypeScript(
+                                    filter_script,
+                                    block_number,
+                                    tx_index,
+                                    io_index,
+                                    match io_type {
+                                        CellType::Input => indexer::CellType::Input,
+                                        CellType::Output => indexer::CellType::Output,
+                                    },
                                 )
-                                .expect("get TxTypeScript should be OK")?;
-                        }
+                                .into_vec(),
+                            )
+                            .expect("get TxTypeScript should be OK")
+                            .is_some(),
+                    };
+                    if !filter_script_matched {
+                        continue;
                     }
                 }
 
                 if let Some([r0, r1]) = filter_block_range {
                     if block_number < r0 || block_number >= r1 {
-                        return None;
+                        continue;
                     }
                 }
 
-                last_key = key.to_vec();
-                Some(Tx {
-                    tx_hash: tx_hash.unpack(),
-                    block_number: block_number.into(),
-                    tx_index: tx_index.into(),
-                    io_index: io_index.into(),
-                    io_type,
-                })
+                let last_tx_hash_is_same = tx_with_cells
+                    .last_mut()
+                    .map(|last| {
+                        if last.tx_hash == tx_hash {
+                            last.cells.push((io_type.clone(), io_index.into()));
+                            true
+                        } else {
+                            false
+                        }
+                    })
+                    .unwrap_or_default();
+
+                if !last_tx_hash_is_same {
+                    tx_with_cells.push(TxWithCells {
+                        tx_hash,
+                        block_number: block_number.into(),
+                        tx_index: tx_index.into(),
+                        cells: vec![(io_type, io_index.into())],
+                    });
+                }
+            }
+
+            Ok(Pagination {
```

Comment:
Does pagination guarantee that all cells belonging to the same tx will return together but not into different pages?

Reply:
Yes, here is the checking:
```
                if tx_with_cells.len() == limit
                    && tx_with_cells.last_mut().unwrap().tx_hash != tx_hash
                {
                    break;
                }
```